rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,270 @@
+related-issues: https://github.com/NixOS/nixpkgs/pull/211832
+---
+
+# Summary
+[summary]: #summary
+
+Auto-generate trivial top-level attribute definitions in `pkgs/top-level/all-packages.nix` from a directory structure that matches the attribute name.
+This makes it much easier to contribute new packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- It is not obvious to package contributors where to add files or which ones to edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in all-packages.nix should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in all-packages.nix to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even locks up editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC consists of two parts, each of which is implemented with a PR to Nixpkgs.
+
+## PR 1: The unit directory standard
+
+This part establishes the new _unit directory standard_ in Nixpkgs.
+This standard is internal to Nixpkgs and not exposed as public interface.
+This standard must be documented in the Nixpkgs manual.
+
+### File structure
+
+Create the initially-empty directory `pkgs/unit`, called _unit base directory_, in Nixpkgs.
+
+Check the following using CI:
+- The unit base directory must only contain subdirectories of the form `pkgs/unit/${shard}/${name}`, called _unit directories_.
+- `name` is a string only consisting of the ASCII characters a-z, A-Z, 0-9, `-` or `_`.
+- `shard` is the lowercased first two letters of `name`, expressed in Nix: `shard = toLower (substring 0 2 name)`.
+- Each unit directory must contain a `package.nix` file and may contain arbitrary other files.
+
+### Semantics
+
+Introduce code to automatically define `pkgs.${name}` for each unit directory as a value equivalent to
+```nix
+pkgs.callPackage pkgs/unit/${shard}/${name}/package.nix { }`
```

Comment:
Glad to see the use of non-empty arguments to `callPackage` going away.

> users viewing a unit directory can be sure that ... the functions arguments correspond directly to attributes in `pkgs`, no exceptions

Don't the `xyzSupport`/`withXyz` booleans conflict with this?

I think "can be sure that functions arguments *without a default value* correspond directly to attributes in `pkgs`" is true though.

Reply:
Oh good point, I fully agree.

It gets a bit more interesting though, because defaults get overridden when packages of the same name do exist. Relevant is [this PR of mine](https://github.com/NixOS/nixpkgs/pull/131271) which would enforce attributes with defaults to not exist (so the default applies).

Reply:
> I also removed the exception

No, you are introducing an exception, by disallowing something.

Custom arguments are needed _at times_, and if something causes you to need one, you will have no option but to move your package out of `unit` and into the legacy hierarchy. When the problem has been resolved, you could move it back, but why even bother. `unit` has failed you for reasons outside your control.

> It would also be weird just to allow this one exception of `callPackage` with custom arguments, when we could also make an exception for custom `callPackage`'s `python3Packages.callPackage`, `haskellPackages.callPackage`, or generally any expression referencing the file.

It is not because the need for those exceptions doesn't vary over time; only by package.

We can easily tighten up the rules when we actually have the opportunity to do so - when we have a good alternative way to provide default arguments, but until then this restriction is counterproductive.



> And the user doesn't have to look at `all-packages.nix`, which is the important part.

If this is a goal, we should not have packages as functions in the first place. Without a better default arguments store, this is not a realistic goal.

Improving the default arguments in package functions is out of scope for this RFC. No _package behavior_ should be affected by this RFC.

Reply:
I agree that requiring no custom arguments is not feasible right now, because it would require moving the package in and out of the folder when the arguments change, which occasionally happens.

But if we allow custom arguments to packages in the `unit` directory, how do we combine this with the ones that are automatically packaged? Does the automation check whether the attribute already exists and then skip it because it has been manually called? Or do we automatically call everything and the manual call does an additional `.override` onto the automatically called package?

Reply:
> It is not because the need for those exceptions doesn't vary over time; only by package.
>
> We can easily tighten up the rules when we actually have the opportunity to do so - when we have a good alternative way to provide default arguments

Those are good points and probably a more general principle to follow, I'll take another think about this.

@piegamesde That's not a problem, we can auto-generate the definitions for all unit directories and just further override the ones with custom arguments. That's how https://github.com/NixOS/nixpkgs/pull/211832 already works.

Reply:
`.override` is only possible when the package evaluates with the bad defaults. This makes `callPackage ../unit/...../packages.nix` a more general solution.

I also think it's conceptually simpler that way, but that's subjective because it does rely on the layout. We might abstract the layout as an `attrsOf path` value somewhere, but that's technically unnecessary as well.

Reply:
I reverted this now, non-empty arguments will be allowed in `all-packages.nix`